src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs

```diff
@@ -27,6 +27,11 @@ public async Task<Organisation> GetByIdAsync(Guid id)
             return await context.Organisations.FirstOrDefaultAsync(org => org.OrganisationId == id);
         }
 
+        public async Task<Organisation> GetByIdWithRelatedOrganisationsAsync(Guid id)
+        {
+            return await context.Organisations.Include(o => o.RelatedOrganisations).ThenInclude(o => o.ChildOrganisation).FirstOrDefaultAsync(org => org.OrganisationId == id);
+        }
+
```

Comment:
To simplify the `OrganisationController.GetUnrelatedOrganisationsAsync` method I'm wondering if it might be worth introducing an additional method, something like `Task<IEnumerable<Organisation>> GetUnrelatedOrganisationsAsync(Guid id)`?

I'm just thinking it might make the controller method simpler if we offloaded most of the work to DB, as if we were writing a query similar to this:

```sql
SELECT * -- Just for illustration!
  FROM dbo.Organisations AS o
 WHERE o.OrganisationId <> @id
   AND o.OrganisationId NOT IN
       (SELECT r.RelatedOrganisationId FROM dbo.RelatedOrganisations AS r WHERE r.OrganisationId = @id);
```

If we add a `RelatedOrganisations` property to the `ApplicationDbContext` I think we can probably do something like this:

```c#
return await context.Organisations
	.Where(o => o.OrganisationId != id)
	.Where(o => !context.RelatedOrganisations.Where(r => r.OrganisationId == id).Select(r => r.RelatedOrganisationId).Contains(o.OrganisationId))
	.ToListAsync();
```

I've not spent a huge amount of time thinking about it so there may be a more elegant way of doing it.

We'd probably want to switch to using `GetByIdAsync` for the first part of the controller method to check the organisation exists.

Admittedly we will have 2 DB calls with this approach but I don't think it's a problem for how frequently this is going to be invoked.

I guess there may even be a way to get it down to 1 DB call but I'll leave that for you to think about (as I don't have time)...

What do you think? Hopefully you'll be able to translate my waffle! 😃

Reply:
i can translate :D i did think about doing this, but theres examples of this work being done in the controller, so took that as the leader. however, i think this function should suffice.

`        public async Task<IEnumerable<Organisation>> GetUnrelatedOrganisations(Organisation organisation)
        {
            return await context.Organisations
                .Where(o => o.OrganisationId != organisation.OrganisationId)
                .Where(o => !organisation.RelatedOrganisations.Any(ro => o.OrganisationId == ro.RelatedOrganisationId))
                .OrderBy(o => o.Name)
                .ToListAsync();
        }`
        
        technically we can't get around the 2 DB calls. if we did a call that gets unrelated on an non-existent org, we would get back a blank list, which would be the same if the org was related to every other org.
        
        so the first call should be to check if the org you're calling exists, then we can call the function above on that org to  get unrelated.
        
        seem ok to you?

Reply:
Yep, seems OK to me! 😃